src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs

```diff
@@ -203,6 +176,60 @@ private async Task CreateBlocksAsync()
             }
         }
 
+        protected virtual async Task WaitBeforeCanMineAsync(int delayMs, CancellationToken cancellation = default(CancellationToken))
+        {
+            await Task.Delay(delayMs, cancellation).ConfigureAwait(false);
+        }
+
+        protected async Task<ChainedHeader> MineBlockAtTimestampAsync(uint timestamp)
+        {
+            ChainedHeader tip = this.consensusManager.Tip;
+
+            Script walletScriptPubKey = this.GetScriptPubKeyFromWallet();
+
+            if (walletScriptPubKey == null)
+            {
+                this.logger.LogWarning("Miner wasn't able to get address from the wallet! You will not receive any rewards.");
+                walletScriptPubKey = new Script();
+            }
+
+            BlockTemplate blockTemplate = this.blockDefinition.Build(tip, walletScriptPubKey);
+
+            blockTemplate.Block.Header.Time = timestamp;
+
+            // Timestamp should always be greater than prev one.
+            if (blockTemplate.Block.Header.Time <= tip.Header.Time)
+            {
+                // Can happen only when target spacing had crazy low value or key was compromised and someone is mining with our key.
+                this.logger.LogWarning("Somehow another block was connected with greater timestamp. Dropping current block.");
+                return null;
+            }
+
+            // Update merkle root.
+            blockTemplate.Block.UpdateMerkleRoot();
+
+            // Sign block with our private key.
+            var header = blockTemplate.Block.Header as PoABlockHeader;
+            this.poaHeaderValidator.Sign(this.federationManager.FederationMemberKey, header);
+
+            ChainedHeader chainedHeader = await this.consensusManager.BlockMinedAsync(blockTemplate.Block).ConfigureAwait(false);
+
+            if (chainedHeader == null)
+            {
+                // Block wasn't accepted because we already connected block from the network.
+                return null;
+            }
+
+            ValidationContext result = this.integrityValidator.VerifyBlockIntegrity(chainedHeader, blockTemplate.Block);
```

Comment:
An integrity check here? is this correct?

Reply:
Yeah. Just to make sure we don't violate integrity. 
It's more like a sanity check that I've added after noticing that I'm not updating merkle tree root but all the validators pass

Reply:
Also if we remove it it won't harm. It's just and additional sanity check which we can afford because integrity rules are pretty cheap